sda-download/api/sda/sda.go

```diff
@@ -450,113 +420,326 @@ func Download(c *gin.Context) {
+	// We had a header specified, so we're done, we shouldn't parse query
+	// parameters
+
+	return
+}
+
+func getCoordinatesFromQuery(c *gin.Context, fileDetails *database.FileDownload) (coords []coordinates, err error) {
+	// Get query params
+	qStart, startGiven := c.GetQuery("startCoordinate")
+	qEnd, endGiven := c.GetQuery("endCoordinate")
+
+	// Neither specified, bail out
+	if !startGiven && !endGiven {
+		// No coordinates specified
+		return
 	}
 
-	// Adapt end coordinate to follow the crypt4gh block boundaries
-	headlength := bytes.NewReader(fileDetails.Header)
-	bodyEnd := int64(fileDetails.ArchiveSize)
-	if end > 0 {
-		var packageSize float64 = 65564 // 64KiB+28, 28 is for chacha20_ietf_poly1305
-		togo := end - start
-		bodysize := math.Max(float64(togo-headlength.Size()), 0)
-		endCoord := packageSize * math.Ceil(bodysize/packageSize)
-		bodyEnd = int64(math.Min(float64(bodyEnd), endCoord))
+	//	Prefill with size from header
+	coord := coordinates{end: int64(fileDetails.DecryptedSize)}
+
+	// Parse and verify coordinates are valid
+	if startGiven {
+		var startValue int64
+		startValue, err = strconv.ParseInt(strings.TrimSpace(qStart), 10, 0)
+		if err != nil {
+			log.Errorf("failed to convert start coordinate %v to integer, %s", qStart, err)
+
+			return
+		}
+
+		if startValue >= 0 {
+			coord.start = startValue
+		} else {
+			// Negative start value, so we want to start from the end
+			coord.start = int64(fileDetails.DecryptedSize) + startValue
+		}
+
+		coord.startSpecified = true
+	}
+
+	if endGiven {
+		var endValue int64
+		endValue, err = strconv.ParseInt(strings.TrimSpace(qEnd), 10, 0)
+		if err != nil {
+			log.Errorf("failed to convert end coordinate %v to integer, %s", qEnd, err)
+
+			return
+		}
+
+		if endValue >= 0 {
+			coord.end = endValue
+		} else {
+			// Negative start value, so we want to start from the end
+			coord.end = int64(fileDetails.DecryptedSize) + endValue
+		}
+
+		coord.endSpecified = true
 	}
 
-	return start, headlength.Size() + bodyEnd, nil
+	if coord.end < coord.start {
+		log.Errorf("endCoordinate=%d must be greater than startCoordinate=%d", coord.end, coord.start)
+		err = fmt.Errorf("endCoordinate must be greater than startCoordinate")
+
+		return
+	}
+
+	if coord.start >= int64(fileDetails.DecryptedSize) || coord.end > int64(fileDetails.DecryptedSize) {
+		log.Errorf("Start (%d) or end (%d) coordinate outside file ", coord.start, coord.end)
+		err = fmt.Errorf("start or end coordinate outside file size")
+
+		return
+	}
+
+	coords = append(coords, coord)
+
+	return
+}
+
+// Calculates the start and end coordinates to use for encrypted transfer.
+// This includes transforming the coordinates from the virtual decrypted stream
+// to the in file encrypted stream (which includes MACs and other overheads)
+// returns a dataeditlist and any possible error
+var adjustCoordinates = func(coords []coordinates, fileDetails *database.FileDownload) (del []uint64, err error) {
```

Comment:
the coordinates should only be adjusted if they are read from the params. If they are read from the `Range: bytes=...` header, the should be left as is. Otherwise the ranges from htsget will not work correctly. For example, htsget usually specifies a request for `Range: bytes=16-123`. A request for this range should give a file with only contain these 108 bytes.